src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,145 @@
+    {
+        private sealed class Analyzer : CSharpSyntaxVisitor<AnalysisResult>
+        {
+            private Analyzer()
+            {
+            }
+
+            public static bool CanConvertToSwitchExpression(SyntaxNode node)
+            {
+                return new Analyzer().Visit(node).Success;
+            }
+
+            private static bool IsDefaultSwitchLabel(SwitchLabelSyntax node)
+            {
+                // default:
+                if (node.IsKind(SyntaxKind.DefaultSwitchLabel))
+                {
+                    return true;
+                }
+
+                if (node.IsKind(SyntaxKind.CasePatternSwitchLabel, out CasePatternSwitchLabelSyntax @case))
+                {
+                    // case _:
+                    if (@case.Pattern.IsKind(SyntaxKind.DiscardPattern))
+                    {
+                        return true;
+                    }
+
+                    // case var _:
+                    // case var x:
+                    if (@case.Pattern.IsKind(SyntaxKind.VarPattern, out VarPatternSyntax varPattern) && 
+                        varPattern.Designation.IsKind(SyntaxKind.DiscardDesignation, SyntaxKind.SingleVariableDesignation))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            public override AnalysisResult VisitSwitchStatement(SwitchStatementSyntax node)
+            {
+                // Fail if the switch statement is empty or any of sections have more than one "case" label.
+                // Once we have "or" patterns, we can relax this to accept multi-case sections.
+                var sections = node.Sections;
+                if (sections.Count == 0 || sections.Any(section => section.Labels.Count != 1))
+                {
+                    return AnalysisResult.Failure;
+                }
+
+                return AnalysisResult.Match(
+                    // If there's no "default" case, we look at the next statement.
+                    // For instance, it could be a "return" statement which we'll use 
+                    // as the default case in the switch expression.
+                    sections.Any(section => IsDefaultSwitchLabel(section.Labels[0]))
+                        ? AnalysisResult.Neutral
+                        : AnalyzeNextStatement(node.GetNextStatement()),
+                    // Iterate over all sections and match section bodies to
+                    // see if they are all convertible to a switch arm's expression.
+                    Aggregate(sections, (result, section) => AnalysisResult.Match(result, AnalyzeSwitchSection(section))));
+            }
+
+            private AnalysisResult AnalyzeNextStatement(StatementSyntax nextStatement)
+            {
+                // Only the following "throw" and "return" can be moved into the switch expression.
```

Comment:
>"return" [](start = 50, length = 8)

Maybe `yield return` too? (that's not blocking this PR though)

Reply:
or, anything, if it's hte last statement in the method? (though i admit this is getting pretty esoteric) :)

Reply:
yield return doesn't make the next statement unreachable. consider this:
```cs
switch (e) {
  case 1:
     yield return 1;
}
yield return 2;
```
it would be incorrect to offer the fix here.

Reply:
Note: we did make this work for "convert if statement to ternary" :)  but i agree, it's a pretty esoteric case.  only support if simple and you want to.

Reply:
For clarification, here's the scenario I had in mind:
```C#
switch (e)
{
  case 1:
    yield return 1;
  default:
    yield return 2;
}
...
```
Which could become:
```C#
yield return e switch { 1 => 1, _ => 2 };
...
```

Reply:
@jcouv That code is illegal right?  You'd still need 'breaks' in the cases so they didn't fall through?

This would also be a nice-to-have.  but only if simple to add IMO.

Reply:
Assuming you have break/yield break, we'll need to actually *generate* `yield break` after the switch in certain circumstances. So it'll be a little more involved than others.

Reply:
Up to you :)   I personally could totally live without this.  I searched roslyn and couldn't find us doing this anywhere, so it's much less important to me.  Just as long as you don't *break* code that does this, tha'ts what matters.

Note: we ended up fixing this for `convert 'if'` because i was using IOperation, and i didn't realize that IOp merged 'return' and 'yield-return' into the same IOp.  But, that doesn't apply to you since you're syntax-based.  So i have no issues at all with this not being done :)

Reply:
> That code is illegal right? You'd still need 'breaks'
Good point. Yes.

As I said, I'm taking back this suggestion. I think it's a much smaller case and probably not worthwhile.